heron/scheduler-core/src/java/com/twitter/heron/scheduler/RuntimeManagerMain.java

```diff
@@ -347,14 +347,18 @@ protected boolean validateRuntimeManage(
       return false;
     }
 
-    // Check whether cluster/role/environ matched
-    ExecutionEnvironment.ExecutionState executionState = adaptor.getExecutionState(topologyName);
-    if (executionState == null
-        || !executionState.getCluster().equals(Context.cluster(config))
-        || !executionState.getRole().equals(Context.role(config))
-        || !executionState.getEnviron().equals(Context.environ(config))) {
-      LOG.severe("cluster/role/environ not matched");
-      return false;
+    // Check whether cluster/role/environ matched. It is possible to get into a state where some of
+    // the data does not actually exist in zookeeper. If we are in this state them assume the
+    // runtime is valid.
+    if (adaptor.doesExecutionStateExist(topologyName)) {
```

Comment:
Yeah. It doesn't feel great to tell people they have to restart their topology in order to do some commands. In the end its probably not the end of the world but feels like there might be a different change to automatically recover from this scenario. The nice part about possibly having the tmaster manager that state is that it can be restored because its encoded in the arguments or in a protobuf object somewhere that could be re-used to automatically recover and restore the missing state.

That being said I think we can make an incremental improvement here and think more about some sort of recovery mechanism.